python/mlir_structured/dialects/indexing.py

```diff
@@ -29,90 +46,199 @@
+    }[py_val.dtype.type]
+  else:
+    raise NotImplementedError(
+        f"Unsupported Python value {py_val=} with type {type(py_val)}")
+
+
+def constant(
+    cst: Union[int, float, bool, np.ndarray],
+    type: Optional[Type] = None,
+    index: Optional[bool] = None,
+) -> arith.ConstantOp:
+  assert isinstance(cst, (int, float, bool, np.ndarray))
+  if index is not None and index:
+    type = IndexType.get()
+  elif type is None:
+    type = infer_mlir_type(cst)
+  assert type is not None
+
+  if isinstance(cst, np.ndarray):
+    type = RankedTensorType.get(cst.shape, type)
+    cst = DenseElementsAttr.get(
+        cst,
+        type=type,
+    )
+
+  cst = arith.ConstantOp(type, cst)
+
+  return cst
+
+
+# TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
+class ArithValueMeta(type(Value)):
+
+  def __call__(cls, *args, **kwargs):
+    if len(args) != 1:
+      raise ValueError("Only one non-kw arg supported.")
+    arg = args[0]
+    if isinstance(arg, (OpView, Operation, Value)):
+      if isinstance(arg, (Operation, OpView)):
+        assert len(arg.results) == 1
+      val = get_op_result_or_value(arg)
+      arg_copy = None
+    elif isinstance(arg, (int, float, bool, np.ndarray)):
+      dtype = kwargs.get("dtype")
+      index = kwargs.get("index")
+      arg_copy = deepcopy(arg)
+      val = constant(arg, dtype, index).result
+    else:
+      raise NotImplementedError(
+          f"{cls.__name__} doesn't support wrapping {arg}.")
+
+    cls_obj = cls.__new__(cls, val)
+    cls.__init__(cls_obj, val, arg_copy)
+    return cls_obj
+
+
+class ArithValue(metaclass=ArithValueMeta):
+
+  def __new__(cls, val):
+    # The call chain after this is the __init__ here *and then* Value.__init__
+    return super().__new__(cls, val)
+
+  def __init__(self,
+               val,
+               arg: Optional[Union[int, float, bool, np.ndarray]] = None):
+    self.__arg = arg
+    super().__init__(val)
+
+  @lru_cache(maxsize=1)
+  def is_constant(self) -> bool:
+    return isinstance(self.owner.opview, arith.ConstantOp)
+
+  @lru_cache(maxsize=1)
   def __str__(self):
     s = res_val_reg.findall(super().__str__())
     assert len(s) == 1
-    return f"Scalar({s[0]}, {self.type})"
+    if self.is_constant():
+      v = str(self.literal_value)
+      v = f"{v[:10]}..." if len(v) > 10 else v
+      return f"{self.__class__.__name__}({s[0]}, {self.type}, {v})"
+    else:
+      return f"{self.__class__.__name__}({s[0]}, {self.type})"
+
+  @lru_cache(maxsize=1)
+  def __repr__(self):
+    return str(self)
+
+  def __binary_op(self, other: "Tensor", op: str) -> "Tensor":
+    assert op in {"add", "sub", "mul"}
+    if self.type != other.type:
+      raise ValueError(f"{self=} {other=} must have the same type.")
+
+    if self.is_constant() and other.is_constant():
+      lhs, rhs = self.literal_value, other.literal_value
+      op = operator.attrgetter(op)(operator)
```

Comment:
This is critical - effectively "constant folding at runtime" - for index arithmetic.

Reply:
Same again: please document in the code.